### PR TITLE
adding event rankings

### PR DESCRIPTION
**Pull request by rytonbay**

### CODE REVIEW

File: TBAKit/Classes/TBAEvent.swift [hunk 1]
```diff
@@ -186,6 +191,40 @@ public struct TBAEvent: TBAModel {
...
 
+public struct TBAEventRanking: TBAModel {
+    
+    public var teamKey: String
+    public var rank: Int
+    public var qualAverage: Double?
+    public var record: [String: Int]
+    public var tieBreakers: [Int]
+    
+    public init?(json: [String: Any]) {
```
Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: Let's make sure we're parsing everything here... looks like there's things like `matches_played`, `extra_stats`, and `dq` don't get parsed

File: TBAKit/Classes/TBAEvent.swift [hunk 2]
```diff
@@ -186,6 +191,40 @@ public struct TBAEvent: TBAModel {
 
 }
 
+public struct TBAEventRanking: TBAModel {
+    
+    public var teamKey: String
+    public var rank: Int
+    public var qualAverage: Double?
+    public var record: [String: Int]
```
Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: WLT is actually an object in the API... maybe we can add some type safety around `record`? Break it in to 3 individual properties - win, loss, and tie? 
screen shot 2017-06-04 at 11 01 17 am

File: TBAKit/Classes/TBAEvent.swift [hunk 3]
```diff
@@ -186,6 +191,40 @@ public struct TBAEvent: TBAModel {
...
+        guard let teamKey = json["team_key"] as? String else {
+            return nil
+        }
+        self.teamKey = teamKey
+        
+        guard let rank = json["rank"] as? Int else {
+            return nil
+        }
+        self.rank = rank
+        if let qualAverage = json["qual_average"] as? Double {
```
Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: Can we get some newlines around here? Kinda visually separate things?

File: TBAKit/Classes/TBAEvent.swift [hunk 4]
```diff
@@ -153,17 +153,22 @@ public struct TBAEvent: TBAModel {
...
+//                completion(array, nil)
+//            } else {
+//                completion(nil, APIError.error("Invalid JSON"))
+//            }
+//        }
+//    }
     
+    public static func fetchRankings(key: String, completion: @escaping ([String: Any]?, Error?) -> ()) -> URLSessionDataTask {
+        let method = "event/\(key)/rankings"
+        return TBAKit.sharedKit.callDictionaryMethod(method, completionHandler: completion)
```
Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: So Rankings is actually an interesting endpoint... if we have the model inside of TBAKit, we don't want to be instantiating those models outside of TBAKit. Maybe we can take another look at how we're getting this data and try to parse/return an array of `TBAEventRanking` and the `sort_order_info` in the completion block

File: TBAKit/Classes/TBAEvent.swift [hunk 5]
```diff
@@ -186,6 +191,40 @@ public struct TBAEvent: TBAModel {
...
+        self.teamKey = teamKey
+        
+        guard let rank = json["rank"] as? Int else {
+            return nil
+        }
+        self.rank = rank
+        if let qualAverage = json["qual_average"] as? Double {
+            self.qualAverage = qualAverage
+        }
+        guard let record = json["record"] as? [String: Int] else {
```
Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: A `null` record is actually an acceptable response here... see 2015, when there were no "winners" and "losers" https://www.thebluealliance.com/api/v3/event/2015miket/rankings

We need to change record to be optional

File: TBAKit/Classes/TBAEvent.swift (hunk 4 above)
Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: Keep in mind when building this one that not every event actually has rankings info

Comment (ZachOrr) [inline TBAKit/Classes/TBAEvent.swift]: Check out the approach I took for district points, which works the same way - I think that's what we're going to have to do for this one https://github.com/ZachOrr/TBAKit/commit/13d5a655757242d7e09addc29802f9f97863db72

File: TBAKit/Classes/TBAEvent.swift [hunk 6]
```diff
@@ -153,17 +153,22 @@ public struct TBAEvent: TBAModel {
...
-        let method = "event/\(eventKey)/district_points"
-        return TBAKit.sharedKit.callArrayMethod(method) { (array, error) in
-            if let array = array as? [[String: String]] {
-                completion(array, nil)
-            } else {
-                completion(nil, APIError.error("Invalid JSON"))
-            }
-        }
-    }
+//    public static func fetchRankings(_ eventKey: String, completion: @escaping (_ alliances: [[String: String]]?, _ error: Error?) -> ()) -> URLSessionDataTask {
```
Comment (rytonbay) [inline TBAKit/Classes/TBAEvent.swift]: Meant to do that, sorry.

**Comment by ZachOrr** [comment]
Hey Bryton - going to close this one. I pushed a huge new update to TBAKit which built out all models/endpoints for the existing v3 API. Updating TBA now... please build off that moving forward :)

File: TBAKit/Classes/TBAEvent.swift (hunk 4 above)
Comment (rytonbay) [inline TBAKit/Classes/TBAEvent.swift]: Yeah, that makes a ton of sense! I love your codebase, it's beautiful, and that definitely gives me a motivation to keep it that way. 

What do we want the `sort_order_info` for again? It's just looking like this where I'm seeing it: 
```
"sort_order_info": [
    {
      "name": "string",
      "precision": 0
    }
  ]
```

File: TBAKit/Classes/TBAEvent.swift (hunk 2 above)
Comment (rytonbay) [inline TBAKit/Classes/TBAEvent.swift]: As in 3 properties of `TBAEventRanking`? Also, where did that screenshot come from?

File: TBAKit/Classes/TBAEvent.swift (hunk 1 above)
Comment (rytonbay) [inline TBAKit/Classes/TBAEvent.swift]: That is true, and it makes sense to pass everything. Will do.

File: TBAKit/Classes/TBAEvent.swift (hunk 2 above)
Comment (rytonbay) [inline TBAKit/Classes/TBAEvent.swift]: Never-mind, found how to see the model descriptions like that.

File: TBAKit/Classes/TBAEvent.swift (hunk 3 above)
Comment (rytonbay) [inline TBAKit/Classes/TBAEvent.swift]: Incidentally just noticed and fixed that while making the other changes.

**Comment by rytonbay** [comment]
Still got to learn from this because I was making the requested changes when you closed it. :) Sorry if you get spammed with my now-irrelevant replies.

**Pull request closed without merging**